Foundation/NSValue.swift

```diff
@@ -157,11 +157,12 @@ open class NSValue : NSObject, NSCopying, NSSecureCoding, NSCoding {
 extension NSValue : _Factory {}
 
 internal protocol _Factory {
-    init(factory: () -> Self)
+    associatedtype StaticSelf = Self
+    init(factory: () -> StaticSelf)
 }
 
 extension _Factory {
-    init(factory: () -> Self) {
-        self = factory()
+    init(factory: () -> StaticSelf) {
+        self = factory() as! Self
```

Comment:
The as! might be problematic with the CF changes I have for generalizing the factory patterns. I know I had a problem with unsafeReferenceCast which is approx the same as as!

Reply:
What kind of problems were you having?

Reply:
Can you try this patch with your in-progress changes and let me know if you run into problems? It should work with -swift-version 4 as well.